Another redis implementation

euri10 opened the pull request:
fixes https://github.com/abersheeran/asgi-ratelimit/issues/8

ok this is a WIP but I think it is a nice implementation, there is still some work to polish it (~~most importantly is the fact that the block feature happens to works by accident I think~~, but I have the feeling it can get added to the script easily), remove debug stuff etc but I wanted to show it to maybe talk with concrete examples about it.

- [x] fix block
- [ ] blocks in the script directly ?
- [x] remove debug stuff

the key concept here is that there is just one script, that avoids calling redis twice per round-trip (one for increase limits, one for decrease) so it should be faster than the current implementation.

it implements a sliding window with sorted sets : I've been inspired by https://levelup.gitconnected.com/implementing-a-sliding-log-rate-limiter-with-redis-and-golang-79db8a297b9e that does the same but only for a single time window, this one is adapted to make it possible to run multiple independent sliding windows

let me know if that's something worth continuing. note also the failing tests are on the current redis implementation because of the `/multiple` tests added, it passes fine on the new implementation

abersheeran commented on ratelimit/backends/base.py:
```diff
@@ -75,14 +75,6 @@ async def set_block_time(self, user: str, block_time: int) -> None:
-            await self.set_block_time(user, rule.block_time)
-
-        return allow
+        raise NotImplementedError()
```
It seems that there is no need to modify it, this method can be overridden in the subclass.

abersheeran commented on ratelimit/backends/redis.py:
```diff
@@ -82,3 +82,15 @@ async def set_block_time(self, user: str, block_time: int) -> None:
     async def is_blocking(self, user: str) -> bool:
         return bool(await self._redis.get(f"blocking:{user}"))
+
+    async def allow_request(self, path: str, user: str, rule: Rule) -> bool:
```
As mentioned above.

abersheeran commented:
Looks like a good change!

euri10 commented on ratelimit/backends/slidingredis.py:
```diff
@@ -0,0 +1,110 @@
+from . import BaseBackend
+
+
+class TimeFilter(logging.Filter):
+    def filter(self, record):
+        try:
+            last = self.last
+        except AttributeError:
+            last = record.relativeCreated
+        delta = datetime.datetime.fromtimestamp(
+            record.relativeCreated / 1000.0
+        ) - datetime.datetime.fromtimestamp(last / 1000.0)
+        record.relative = "{0:.2f}".format(
+            delta.seconds + delta.microseconds / 1000000.0
+        )
+        self.last = record.relativeCreated
+        return True
+
+
+logger = logging.getLogger("ratelimit")
+logger.setLevel(logging.DEBUG)
+ch = logging.StreamHandler()
+ch.setLevel(logging.DEBUG)
+formatter = logging.Formatter(
+    fmt="+%(relative)s - %(name)s - %(levelname)s - %(message)s"
+)
+logger.addHandler(ch)
+[hndl.addFilter(TimeFilter()) for hndl in logger.handlers]
+[hndl.setFormatter(formatter) for hndl in logger.handlers]
```
what to do with this, comment it here or paste it in the PR so that it is useful to someone else down the road. It spits out a nice logging output so that you can see time spent between 2 log events, this avoids headaches when trying to figure out if yes/no a request should be blocked:
```
+0.00 - ratelimit - DEBUG - 1616057360.867704 [0, 2] : False
+1.01 - ratelimit - DEBUG - 1616057361.8722386 [1, 1] : True
+0.00 - ratelimit - DEBUG - 1616057361.876414 [0, 0] : False
+2.01 - ratelimit - DEBUG - 1616057363.8830643 [1, 0] : False
```

abersheeran replied:
Maybe putting it in tests would be a better choice.

euri10 replied:
definitely yes ! I didnt think about it

euri10 commented:
I implemented requested changes, added test for multiple rules only for the `SlidingRedisBackend` so that it does not fail on the `RedisBackend` and commented out the logging, just need to know if you want to keep it commented in the code or pasted here for future reference, it helped me a lot debugging the lua script so it might prove useful to someone elese in the future.

euri10 commented:
I had to poetry update to please the CI that failed in https://github.com/abersheeran/asgi-ratelimit/pull/10/commits/80157a0ebbc631d3d065b43a0ab39ee35f648902

this explains some changes in poetry.lock and black formatting since it upped the version

abersheeran commented:
> I had to poetry update to please the CI that failed in [80157a0](https://github.com/abersheeran/asgi-ratelimit/commit/80157a0ebbc631d3d065b43a0ab39ee35f648902)
> 
> this explains some changes in poetry.lock and black formatting since it upped the version

😀It doesn't matter, this is an occasional issue on GitHub. In fact, re-running CI should bring it back to normal.

abersheeran commented:
Do you still want to modify anything? If not, I will merge this PR.

euri10 commented:
> Do you still want to modify anything? If not, I will merge this PR.

let's roll !
